src/cluster.c

```diff
@@ -7576,82 +1294,175 @@ int clusterRedirectBlockedClientIfNeeded(client *c) {
-        raxRemove(server.cluster->slots_to_channels,indexed,keylen+2,NULL);
+    for (int i = 0; i <= CLUSTER_SLOTS; i++) {
+        /* Find start node and slot id. */
+        if (n == NULL) {
+            if (i == CLUSTER_SLOTS) break;
+            n = getNodeBySlot(i);
+            start = i;
+            continue;
+        }
+
+        /* Add cluster slots info when occur different node with start
+         * or end of slot. */
+        if (i == CLUSTER_SLOTS || n != getNodeBySlot(i)) {
+            addNodeReplyForClusterSlot(c, n, start, i-1);
+            num_masters++;
+            if (i == CLUSTER_SLOTS) break;
+            n = getNodeBySlot(i);
+            start = i;
+        }
     }
-    if (indexed != buf) zfree(indexed);
+    setDeferredArrayLen(c, slot_replylen, num_masters);
 }
 
-void slotToChannelAdd(sds channel) {
-    slotToChannelUpdate(channel,1);
+/* -----------------------------------------------------------------------------
+ * Cluster functions related to serving / redirecting clients
+ * -------------------------------------------------------------------------- */
+
+/* The ASKING command is required after a -ASK redirection.
+ * The client should issue ASKING before to actually send the command to
+ * the target instance. See the Redis Cluster specification for more
+ * information. */
+void askingCommand(client *c) {
+    if (server.cluster_enabled == 0) {
+        addReplyError(c,"This instance has cluster support disabled");
+        return;
+    }
+    c->flags |= CLIENT_ASKING;
+    addReply(c,shared.ok);
 }
 
-void slotToChannelDel(sds channel) {
-    slotToChannelUpdate(channel,0);
+/* The READONLY command is used by clients to enter the read-only mode.
+ * In this mode slaves will not redirect clients as long as clients access
+ * with read-only commands to keys that are served by the slave's master. */
+void readonlyCommand(client *c) {
+    if (server.cluster_enabled == 0) {
+        addReplyError(c,"This instance has cluster support disabled");
+        return;
+    }
+    c->flags |= CLIENT_READONLY;
+    addReply(c,shared.ok);
 }
 
-/* Get the count of the channels for a given slot. */
-unsigned int countChannelsInSlot(unsigned int hashslot) {
-    raxIterator iter;
-    int j = 0;
-    unsigned char indexed[2];
-
-    indexed[0] = (hashslot >> 8) & 0xff;
-    indexed[1] = hashslot & 0xff;
-    raxStart(&iter,server.cluster->slots_to_channels);
-    raxSeek(&iter,">=",indexed,2);
-    while(raxNext(&iter)) {
-        if (iter.key[0] != indexed[0] || iter.key[1] != indexed[1]) break;
-        j++;
+/* The READWRITE command just clears the READONLY command state. */
+void readwriteCommand(client *c) {
+    if (server.cluster_enabled == 0) {
+        addReplyError(c,"This instance has cluster support disabled");
+        return;
     }
-    raxStop(&iter);
-    return j;
+    c->flags &= ~CLIENT_READONLY;
+    addReply(c,shared.ok);
 }
+
+#ifdef REDIS_CLUSTER_FLOTILLA
+#include "cluster_flotilla.c"
+#else
```

Comment:
since this is not part of this PR, i think we should drop this.
and any reference to "Flotilla" from the current PR.
instead, let's add REDIS_CLUSTER_LEGACY that can be turned off.

is it easy to also make the necessary changes so that redis can be compiled and run with `cluster_enabled`, even with the legacy cluster excluded?
i.e. i imagine that it would work with all the slot awareness and restrictions, but without any cluster commands and cluster bus.
if this is complicated, let's leave it for a future PR, but at least in this one, avoid the "flotilla" defines and referring to files that don't exist, and instead prepare the ground for being able to just exclude the legacy cluster in a later PR.

Reply:
Done. No mention of flotilla in the code base at all